File: scripts/inspect_glb.js
import fs from 'fs';
import path from 'path';

const filePath = path.join(process.cwd(), 'client/public/models/plmxs.glb');

try {
  const buffer = fs.readFileSync(filePath);
  
  // GLB Header: Magic (4) + Version (4) + Length (4) = 12 bytes
  const magic = buffer.readUInt32LE(0);
  const version = buffer.readUInt32LE(4);
  const totalLength = buffer.readUInt32LE(8);
  
  if (magic !== 0x46546C67) { // 'glTF'
    console.error('Not a valid GLB file');
    process.exit(1);
  }
  
  // Chunk 0: Length (4) + Type (4) + Data
  const chunkLength = buffer.readUInt32LE(12);
  const chunkType = buffer.readUInt32LE(16);
  
  if (chunkType !== 0x4E4F534A) { // 'JSON'
    console.error('First chunk is not JSON');
    process.exit(1);
  }
  
  const jsonBuffer = buffer.subarray(20, 20 + chunkLength);
  const jsonStr = jsonBuffer.toString('utf8');
  const json = JSON.parse(jsonStr);
  
  console.log('--- Model Analysis ---');
  console.log('Meshes:', json.meshes ? json.meshes.length : 0);
  
  if (json.meshes) {
    console.log('\nMesh Names:');
    json.meshes.forEach((mesh, index) => {
      console.log(`${index}: ${mesh.name}`);
      if (mesh.primitives) {
          mesh.primitives.forEach((p, pi) => {
              console.log(`  - Primitive ${pi}: material=${p.material}, targets=${p.targets ? p.targets.length : 0}`);
              if (p.targets) {
                  // GLTF targets usually match 'extras' or 'weights' names if available in mesh
                  // But pure JSON might not show names easily without looking at mesh.extras.targetNames
                  // We'll just check existence.
              }
          });
      }
      if (mesh.extras && mesh.extras.targetNames) {
          console.log(`    Target Names: ${mesh.extras.targetNames.join(', ')}`);
      }
    });
  }

  if (json.nodes) {
      console.log('\nNode Names (Hierarchy):');
      json.nodes.forEach((node, index) => {
          if (node.mesh !== undefined) {
             console.log(`${index}: ${node.name} (Has Mesh)`);
          } else {
             console.log(`${index}: ${node.name}`);
          }
      });
  }
  
} catch (error) {
  console.error('Error parsing GLB:', error);
}

File: scripts/inspect_glb.py
import json
import struct
import os

def inspect_glb(file_path):
    print(f"Inspecting: {file_path}")
    if not os.path.exists(file_path):
        print("File not found.")
        return

    with open(file_path, 'rb') as f:
        magic = f.read(4)
        if magic != b'glTF':
            print("Not a valid binary glTF file.")
            return
        
        version = struct.unpack('<I', f.read(4))[0]
        length = struct.unpack('<I', f.read(4))[0]
        
        print(f"glTF Version: {version}")
        
        # Read chunks
        while f.tell() < length:
            chunk_length = struct.unpack('<I', f.read(4))[0]
            chunk_type = f.read(4)
            
            if chunk_type == b'JSON':
                json_data = f.read(chunk_length)
                data = json.loads(json_data)
                
                # Inspect meshes for morph targets
                if 'meshes' in data:
                    print(f"Found {len(data['meshes'])} meshes.")
                    for i, mesh in enumerate(data['meshes']):
                        print(f"Mesh {i}: {mesh.get('name', 'Unnamed')}")
                        if 'primitives' in mesh:
                            for prim in mesh['primitives']:
                                if 'targets' in prim:
                                    print(f"  - Has {len(prim['targets'])} Morph Targets")
                                    # Try to find target names in extras or mesh weights
                                    if 'weights' in mesh:
                                        print(f"  - Default Weights: {mesh['weights']}")
                                    
                                    # Names are usually in mesh.extras.targetNames or just implied
                                    if 'extras' in mesh and 'targetNames' in mesh['extras']:
                                        print(f"  - Target Names: {mesh['extras']['targetNames']}")
                                else:
                                    print("  - No Morph Targets found.")
                else:
                    print("No meshes found.")
                    
                break
            else:
                f.seek(chunk_length, 1)

import sys

if __name__ == "__main__":
    path = "models/Blue_Outdoor_Shirt.glb"
    if len(sys.argv) > 1:
        path = sys.argv[1]
    inspect_glb(path)
